nsiqcppstyle_rulemanager.py

```diff
@@ -180,62 +200,154 @@ def ResetRegisteredRules(self):
+            def RunRule(lexer, line, lineNumber)
+                lexer is the lexer used to analyze the source, it points the start token of source.
+                line is the text of the physical line just read
+                lineNumber is the line offset of <line> in the file being processed
+        """
         self.lineRules.append(lineRule)
 
     def AddRule(self, rule):
-        """ Add rule on any token """
+        """
+        Add rule on any token
+        Added rule should be function with following prototype:
+            def RunRule(lexer, contextStack)
+                lexer is the lexer used to analyze the source, it points the start token of source.
+                contextStack is a stack showing the lexical context the token resides in
+        """
         self.rules.append(rule)
 
     def AddTypeNameRule(self, typeNameRule):
-        """ Add rule on any type (class / struct / union / namesapce / enum) """
+        """
+        Add rule on any type (class / struct / union / namespace / enum)
+        Added rule should be function with following prototype:
+            def RunRule(lexer, typeName, typeFullName, decl, contextStack, typeContext)
+                lexer is the lexer used to analyze the source, it points the start token of source.
+                typeName is the name of the C++ type encountered (e.g., "CLASS", "STRUCT")
+                typeFullName is the typed variable name (e.g., the class name)
+                decl is a boolean flag denoting whether the function is a declaration (True) or
+                     definition (False)
+                contextStack is a stack showing the lexical context the token resides in
+                typeContext is the new context entry that will be pushed to the contextStack due to this token
+        """
         self.typeNameRules.append(typeNameRule)
 
     def AddTypeScopeRule(self, typeScopeRule):
-        """ Add rule on the any type definition scope """
+        """
+        Add rule when the token is within a type definition scope
+        Added rule should be function with following prototype:
+            def RunRule(lexer, typeName, typeFullName, decl, contextStack, typeContext)
+                lexer is the lexer used to analyze the source, it points the start token of source.
+                contextStack is a stack showing the lexical context the token resides in
+        """
         self.typeScopeRules.append(typeScopeRule)
 
     def AddFileEndRule(self, fileEndRule):
         """
         Add rule on the file end
-        Added Rule should be function with following prototype "def RunRule(lexer, filename, dirname)"
-        lexer is the lexer used to analyze the source. it points the end token of source.
-        filename is the filename analyzed.
-        dirname is the file directory.
+        Added rule should be function with following prototype:
+            def RunRule(lexer, filename, dirname)

```

Comment:
Since we are using Python 3, let's use typing instead of writing the type in comments

Reply:
Here, the function can be annotated as:
```
from typing import TypeVar
from pathlib import Path

# advanced typing
FileName = TypeVar('FileName', str)  # must be string, but is more than just a string
DirName = TypeVar('DirName', Path)

def func(user_function: Callable[[Lexer, FileName, DirName], ReturnType]):
    pass

# simpler typing
def func(user_function: Callable[[Lexer, str, Path], ReturnType]):
    pass
```

Reply:
I thank you for dragging me into 2022 with the use of Python types :-)

- I defined new types using NewType(), rather than TypeVar().  This is because TypeVar() gave me an error message "TypeError: A single constraint is not allowed" while NewType() works exactly as I wanted.
- I've created a new project file, nsiqcppstyle_types.py, and have placed all new types there.
- All NewType() definitions are accompanied by a comment to explain their purpose
- Files that want to use these new types insert an "import nsiqcppstyle_types" statement.
- As a test, I modified the rule function in RULE_4_1_A_B_use_space_for_indentation.py to map the function arguments to types.  If this is acceptable, we can plan on doing this in all RULE files.

Reply:
Done.